base/compiler/typeinfer.jl

```diff
@@ -129,14 +130,22 @@ function finish!(interp::AbstractInterpreter, caller::InferenceState;
             end
         end
         # n.b. relocatability = isa(inferred_result, String) && inferred_result[end]
-        if !@isdefined edges
-            edges = DebugInfo(result.linfo)
+        if !@isdefined di
+            di = DebugInfo(result.linfo)
+        end
+        istoplevel = caller.linfo.def isa Method # don't add backedges to toplevel method instance
+        if istoplevel
+            edges = empty_edges
+        else
+            # TODO remove the next line once we make staticdata_utils.c able to handle `CodeInstance` edge directly
+            edges = Any[edge isa CodeInstance ? edge.def : edge for edge in caller.edges]
```

Comment:
I think the Julia-side refactoring is complete for now. `edges` is now a list of `CodeInstance` instead of `MethodInstance`. By removing this one line, the C-side implementation will be able to handle recursive edges as a list of `CodeInstance`.
@vtjnash I’d like to request your review and assistance with the C-side refactor at this point, if possible.

---

Below are some concerns and ideas I have so far:

- Currently, we’re generating dummy `CodeInstance` objects with `codeinst_as_edge(...)` to represent const-propped callsite edges. These `dummy_edge::CodeInstance` objects don’t contain much useful information outside of `dummy_edge.edges`, so most of their fields are wasted. So `CodeInstance` might be a bit too heavy as an efficient encoding of the recursive edges, and we might want to consider defining a dedicated object just for recursive edges.

- Maybe related to the points above, the roles of `CodeInstance` and `InferenceResult` have become quite similar. A refactor to merge `InferenceResult` into `CodeInstance` might also be worth considering.

- In places like `_hasmethod_tfunc` and `abstract_applicable`, where backedges are needed for method match, or for inlined call sites without a proper `edge::CodeInstance`, we’re generating dummy `dummy_invoke_edge::CodeInstance`s with `codeinst_as_invoke_edge()`. They are sufficient for representing backedges, but I’m not entirely confident they're appropriate as forward edges (`dummy_invoke_edge.edges` are always empty)

- As a minor improvement for the inlining algorithm, we could potentially improve it to skip `code_cache` lookups by using the proper `edge::CodeInstance` in `CallInfo`, but I haven’t implemented that yet (or in a larger context, this will probably be done anyway as part of addressing JuliaLang/julia#54899).

Reply:
> we’re generating dummy CodeInstance objects with codeinst_as_edge(...) to represent const-propped callsite edges. 

That's pretty similar to our "const return ABI" code instances, right? IIRC, those are also mostly used just for their return type + backedges.

> CodeInstance might be a bit too heavy as an efficient encoding of the recursive edges

Do we not have (the equivalent of) MethodTable edges in the new scheme? Are those just CodeInstances w.r.t. to a wide MethodInstance now?

Reply:
> That's pretty similar to our "const return ABI" code instances, right? IIRC, those are also mostly used just for their return type + backedges.

Yes, that's correct. Those `CodeInstance`s aren’t stored in the cache, and only a few of their fields are actually used.

> Do we not have (the equivalent of) MethodTable edges in the new scheme? Are those just CodeInstances w.r.t. to a wide MethodInstance now?
> 
> **edit:** I suppose abstract calls in general are covered by the "N CodeInstances" encoding of an edge but that sounds potentially inefficient?

What I was referring to is the general design of recursive edges. Currently, a `caller::CodeInstance` has `caller.edges::SimpleVector`, and `caller.edges` consists of `mt::MethodTable` /w `sig` edges or other general `edge::CodeInstance`s. However, some of these `edge::CodeInstance`s are created by `codeinst_as_edge(...)` or `codeinst_as_invoke_edge(...)` as mentioned before, and these edges contain unnecessary fields—this was the point I was trying to make. 

Still, I’m not sure if there’s a more efficient design than the current implementation.